### PR TITLE
fix: prepare deadlock

### LINKED ISSUE #4419: gorm deadlock may occur while initialization config 'SkipDefaultTransaction' set to false(default) and 'PrepareStmt' set to true
Prerequisites:  'PrepareStmt' been set to true while initializing gorm

1. when 'SkipDefaultTransaction' not set or set to false, gorm will add a transaction for non-query operations automatically,
```
func BeginTransaction(db *gorm.DB) {
	if !db.Config.SkipDefaultTransaction {
		if tx := db.Begin(); tx.Error == nil {
			db.Statement.ConnPool = tx.Statement.ConnPool
			db.InstanceSet("gorm:started_transaction", true)
		} else if tx.Error == gorm.ErrInvalidTransaction {
                ...
``` 
and it will get a connection, start transaction, and then exec PreparedStmtDB.ExecContext, in  PreparedStmtDB.ExecContext, gorm will try to get PreparedStmtDB.Mux and then exec the SQL
```
0x55972dbfc7c8	sync.runtime_SemacquireMutex+0x48
0x55972e1e70a9	sync.(*RWMutex).RLock+0x459
0x55972e1e6c73	gorm.io/gorm.(*PreparedStmtDB).prepare+0x23
0x55972e1e78d2	gorm.io/gorm.(*PreparedStmtTX).ExecContext+0x92
0x55972e23fb9c	gorm.io/gorm/callbacks.Create.func1+0x37c
0x55972e1d97b5	gorm.io/gorm.(*processor).Execute+0x225
0x55972e1de18a	gorm.io/gorm.(*DB).Create+0xba
```

2. for query operation, gorm does not start transaction, and in PreparedStmtDB.QueryContext->PreparedStmtDB.prepare, it will get the PreparedStmtDB.Mux and then try to get a connection(stmt, err := conn.PrepareContext(ctx, query))
```
0x55972e191c3d	database/sql.(*DB).conn+0x7ad
0x55972e193009	database/sql.(*DB).prepare+0x59
0x55972e192dca	database/sql.(*DB).PrepareContext+0x9a
0x55972e1e6e70	gorm.io/gorm.(*PreparedStmtDB).prepare+0x220
0x55972e1e74bc	gorm.io/gorm.(*PreparedStmtDB).QueryContext+0x8c
0x55972e2379ea	gorm.io/gorm/callbacks.Query+0xfa
0x55972e1d97b5	gorm.io/gorm.(*processor).Execute+0x225
0x55972e1df12d	gorm.io/gorm.(*DB).Take+0xcd
...
```
so when all the connections in pool are exhausted, scenario 1 waitting for mutex while holding all connections, and scenario 2 waiting for connection while holding mutex, then deadlock occurs

### PR DESCRIPTION
**Pull request by a631807682**
- [x] Do only one thing
- [x] Non breaking API changes
- [x] Tested

### What did this pull request do?

close #4419

### CODE REVIEW

File: prepare_stmt.go [hunk 1]
```diff
@@ -50,23 +50,18 @@ func (db *PreparedStmtDB) prepare(ctx context.Context, conn ConnPool, isTransact
...
 	db.Mux.RUnlock()
 
-	db.Mux.Lock()
-	defer db.Mux.Unlock()
-
-	// double check
-	if stmt, ok := db.Stmts[query]; ok && (!stmt.Transaction || isTransaction) {
-		return stmt, nil
-	} else if ok {
-		go stmt.Close()
```
Comment (a631807682) [inline prepare_stmt.go:60]: I removed double check because it was already checked in read lock.

File: prepare_stmt.go [hunk 2]
```diff
@@ -50,23 +50,18 @@ func (db *PreparedStmtDB) prepare(ctx context.Context, conn ConnPool, isTransact
...
-	}
-
 	stmt, err := conn.PrepareContext(ctx, query)
```
Comment (jinzhu) [inline prepare_stmt.go:53]: looks like need to `Prepare` after the lock?

Comment (a631807682) [reply prepare_stmt.go:53]: mean here?
https://github.com/go-gorm/gorm/pull/5568/files#diff-0b194a262799c7ce6ddec3469dedda37d182b3a97d16f079bc234778aaa6312eR60

Comment (jinzhu) [reply prepare_stmt.go:53]: Yes, change it to something like:

```go
	db.Mux.Lock()
	stmt, err := conn.PrepareContext(ctx, query)
	if err != nil {
		return Stmt{}, err
	}

	cacheStmt := Stmt{Stmt: stmt, Transaction: isTransaction}
	db.Stmts[query] = cacheStmt
	db.PreparedSQL = append(db.PreparedSQL, query)
	db.Mux.Unlock()
```

Comment (a631807682) [reply prepare_stmt.go:53]: Okay I got it, we can't lock conn.PrepareContext, I'll fix the code later and describe why the deadlock happens.

**Comment by a631807682** [comment]
related code
https://github.com/golang/go/blob/master/src/database/sql/sql.go#L528
https://github.com/golang/go/blob/8cb350d69a1b0765c1c81301583d6fd99fb9d74b/src/database/sql/sql.go#L1359

**Comment by YJinHai** [comment]
hi,When will this issue be solved together?

**Pull request merged**